### ISSUE TITLE
Stop relying on SparseM::chol in slm.fit.csr.fixed

### ISSUE THREAD

**Issue by jwbowers** [markmfredrickson/RItools#134]
As of SparseM 1.8.2 our use of `SparseM::chol` in `slm.fit.csr.fixed` no longer works and causes balanceTest to throw errors.

We have been using it to produce OLS coefs for edge cases, for example, where x'x is something like (in non-sparse form):

```
     [,1] [,2]
[1,]   23    0
[2,]    0    0
```

Here is `slm.fit.csr.fixed`:

```
slm.fit.csr.fixed <- function (x, y, ...)
  {
      if (is.matrix(y))
          {
              n <- nrow(y)
              ycol <- ncol(y)
          } else {
              n <- length(y)
              ycol <- 1
          }
      p <- x@dimension[2]
      if (n != x@dimension[1])
          stop("x and y don't match n")
      chol <- SparseM::chol(t(x) %*% x, ...)
      xy <- t(x) %*% y
      coef <- SparseM::backsolve(chol, xy)

      if (is.vector(coef)) {
        coef <- matrix(coef, ncol = ycol, nrow = p)
    }

      fitted <- as.matrix(x %*% coef)
      resid <- y - fitted
      df <- n - p
      list(coefficients = coef, chol = chol, residuals = resid,
          fitted = fitted, df.residual = df)
  }
```

**Comment by jwbowers** [comment]
One simple solution is to use the following within the above function (perhaps now renamed).

```
coef  <- lm.fit(as.matrix(x),y)$coef
coef[is.na(coef)] <- 0
```

Which works with this example:

```
testdata <- nuclearplants
testdata$date[testdata$date < 68] <- NA
balanceTest(pr ~ . - cost + strata(pt), data = testdata)
```

I don't know what kind of performance problems we might have so I am not implementing this as a patch just yet.

**Comment by jwbowers** [comment]
In this particular case the coefficient is just `mean(y[x[,1]==1])` or `sum(x[,1]*y)/sum(x[,1])` (assuming `x` is an ordinary matrix --- this code doesn't work for SparseM x).

**Comment by markmfredrickson** [comment]
Since `t(x) %*% x` is going to be of modest dimension, using the dense version of `lm.fit` is probably not going to be much of a performance hit.

**Comment by benthestatistician** [comment]
I agree that the dimension of `t(x) %*% x`  will be more modest than that of `x`, but I aspire to handle cases in which its dimension is of the same magnitude as `x`'s: with pair matching, the row or column extent of `t(x) %*% x`  can be half of `x`'s row extent. 

In the particular edge case we're seeing here, `slm.fit.csr.fixed` would side-step the pothole by looking for zeroes in `xy` before figuring `chol(t(x) %*% x)`.  Whereever it found zeroes, it pre-decides that the `coef` value will be a zero, then it removes the corresponding rows and columns form `xy` and from $x'x$ (=`t(x) %*% x`=:`xprimex`) before proceding to the cholesky decomposition and backsolve ops.

**Comment by benthestatistician** [comment]
PS: `slm.fit.csr.fixed()` internal docs should be updated to reflect whatever positive definiteness requirement may be in force.

**Comment by benthestatistician** [comment]
(I assigned myself, but I'm hoping @adamrauh will lead the fixing from here.)

**Comment by benthestatistician** [comment]
Update: please don't hesitate to try out the `lm.fit()`-based solution you mention above, @jwbowers, and if it works as hoped to push a maintenance release with that mechanism in it.  In the interests of staying ahead for unwanted CRAN-scrutiny. If the approach I described worked out, we can push another maintenance release with it.

**Comment by jwbowers** [comment]
I have some time today and I'll work on the `lm.fit()` approach in a new branch. Others feel free to think about a sparse approach that scales better than `lm.fit()`.

**Comment by benthestatistician** [comment]
Just noting that when(if) Jake's maintenance fix in #135 gets replaced with a SparseM native fix, we should:

1. Restore a `chol` entry to the `slm.fit.csr.fixed()` return value, if/as necessary to match the output of `slm.fit.csr()`. 
2. Re-enable the test at [test.utils.R#L48-50](https://github.com/markmfredrickson/RItools/blob/5041b4195b15cadbe2d336c0a1a9b32105e53745/tests/testthat/test.utils.R#L48-L50), after updating as appropriate.

**Comment by jwbowers** [comment]
RItools 0.3-4 has been submitted to CRAN using main as of now. FYI.

**Comment by adamrauh** [comment]
Circling back around to this, trying to implement @benthestatistician's suggestion outlined above. I have a few questions. I'll also note that I do not understand what I'm looking at here very well, so bear with me.

Using @jwbowers 's example:
```
testdata <- nuclearplants
testdata$date[testdata$date < 68] <- NA
balanceTest(pr ~ . - cost + strata(pt), data = testdata)
```

Within `slm.fit.csr.fixed()`, create:

```
xy <- t(x) %*% y
xprimex <- t(x) %*% x

Browse[1]> xy
        [,1]
[1,] 1586.86
[2,]    0.00

Browse[1]> as.matrix(xprimex)
     [,1] [,2]
[1,]   23    0
[2,]    0    0

```

> identifying zeroes on the diagonal of  `xprimex` 

Without having to convert to a dense matrix format, we could identify that with `diag(xprimex)`. Easy enough.

> getting rid of the corresponding entries of `xy` 

so, since there's a 0 at (2,2) in `xprimex`, remove the second row (and column if it existed) of `xy`?

> and then Cholesky-decomposing only the diagonal block of `xprimex` with nonzero numbers on the diagonal.

In this example, that would just leave something like: 

```
     [,1] 
[1,]   23  
```
right? I tried manually plugging these modified bits into `SparseM::chol()` and subsequently backsolving, and ended up with an answer that did not match, so I suspect I'm getting some things wrong here. 

Digging around some, it also doesn't appear that SparseM offers a good way to subset an object like this directly. Is there something we can reliably use about the structure of `xprimex` to pull out this data via some kind of matrix multiplication, creation of a new SparseM, etc.?

**Comment by benthestatistician** [comment]
1. You've got the idea of how I think the dimension of `xprimex` ought to be reduced.  With regard to the `xy` of your example, you'd only want to remove the second row.  (Not that I think it will arise that it has multiple columns.
2.  I think you're right that there's a matrix multiplication that reduces the dimension for you.  Writing `diagzeroes` for a Boolean vector indicating which diagonal positions of `xprimex` are zeroes and `k_` for the column dimension of `x`,  with the matrix `mat = diag(nr=k_0)[, !diagzeroes]`  you could do `t(mat) %*% xprimex %*% mat` and `t(mat) %*% xy`.  It'd be better to build up a sparse version of `mat`, though. 
3. Re answers not coming out, not sure what's going on but you might check that the base R Cholesky does what it should.  The Cholesky decomposition of a 1 by 1 matrix is a weird edge case, maybe it that's less well supported in SparseM.

**Comment by jwbowers** [comment]
Closing this issue as solved with [a660143](https://github.com/markmfredrickson/RItools/commit/a6601434718cd3fef1b63d9cf000b5b1848c8872)
